### ISSUE TITLE
Unable update CRD identity information with a reference for this node

### ISSUE THREAD

**Issue by joestringer** [cilium/cilium#9792]
User report from Slack:

> Curious if anyone has seen an error like `Unable update CRD identity information with a reference for this node` in their cilium pod logs? Might be a red herring. I'm running Cilium on EKS in CNI chaining mode. It's been fine for a couple of weeks, but today it degraded to a state where if a pod gets scheduled on a node, it gets stuck in ContainerCreating. In the pod's events, I see errors like:
> `NetworkPlugin cni failed to set up pod "REDACTED" network: unable to create endpoint: Put http:///var/run/cilium/cilium.sock/v1/endpoint/cilium-local:0: context deadline exceeded`.
> If I delete the cilium pod on the node, all other stuck pods get scheduled successfully once the cilium pod comes up again.

> I see a lot of this error as well
> `2019-12-04T23:00:19.707077314Z level=warning msg="Key allocation attempt failed" attempt=11 error="unable to create slave key 'k8s:app=REDACTED;k8s:io.cilium.k8s.policy.cluster=default;k8s:io.cilium.k8s.policy.serviceaccount=REDACTED;k8s:io.kubernetes.pod.namespace=default;k8s:service-type=rest;': Operation cannot be fulfilled on ciliumidentities.cilium.io \"64380\": StorageError: invalid object, Code: 4, Key: /registry/cilium.io/ciliumidentities/64380, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: f6d81d4c-0a28-11ea-b6ee-122bfb3499ba, UID in object meta: " key="[k8s:app=REDACTED k8s:io.cilium.k8s.policy.cluster=default k8s:io.cilium.k8s.policy.serviceaccount=REDACTED k8s:io.kubernetes.pod.namespace=default k8s:service-type=rest]" subsys=allocator`
> The error message `StorageError: invalid object, Code: 4` seems to come from kubernetes/kube-apiserver. I also notice that the crd kvstore implementation uses a cache in front of kube-api. I’m wondering if there could be some cache invalidation issues at play here.

> So we think we know what causes this issue, Amazon periodically upgrades the “eks-version” which includes upgrading the control plane to a newer patch version.
> During this time the control plane is restarted by AWS
> These changes are not rolled out to all EKS clusters at the same, they apparently do some kind of a random rollout process. Today two of our clusters got upgraded from eks.2->eks.6. After some pod churn both clusters starting seeing pods being failed to start.
> We also believe a similar upgrade happened around the time when we first saw this issue a few weeks ago (we think that the version went from 1->2).

**Comment by shelby-moore** [comment]
Let me know if we can help with contributing a fix for this, we'd be happy to do so.

**Comment by raybejjani** [comment]
@shelby-moore First, thank you for figuring out what was happening. This was popping up every so often and it was driving me nuts!
I'm still trying to figure out how to reproduce this (and so fix it). As you pointed out, there are two caches at play here. The only one that makes sense would be a local `k8s.io/client-go/tools/cache.Store` object used in a list-and-watch. We use the object copy in there when updating k8s objects, in this case a `ciliumidentity` instance. The "slave key" in this case is an update to the status field on the ciliumidentity instance named the numeric identity (64380 in the example above) , adding a nodename->timestamp entry to a map.

I'll try to get some eks clusters up where we can watch for this but, since you already have some, would it be possible for you to enable debug on a node experiencing this problem?
As I recall, it tends to persist until cilium-agent (or it's pod) is restarted. To do this on the fly, without restarting the pod, you would need to issue a `cilium config Debug=true` to the pod (via a `kubectl exec`). This causes a bunch of datapath regenerations and a lot more log output. This might break production loads!
Hopefully, we would then see more `subsys=crd-allocator` log lines and that might point to something. Some deep k8s client library may also print something like:
```
Jan 06 16:42:16 k8s1 cilium-agent[16137]: W0106 16:42:16.377422   16137 reflector.go:340] github.com/cilium/cilium/pkg/k8s/watchers/service.go:105: watch of *v1.Service ended with very short watch: github.com/cilium/cilium/pkg/k8s/watchers/service.go:105: Unexpected watch close - watch lasted less than a second and no items received
```

**Comment by shelby-moore** [comment]
Hey @raybejjani - we actually restarted all the Cilium pods on the clusters that were experiencing this, as the issue was impacting other teams. I'll try reproducing on a new cluster tomorrow.

**Comment by shelby-moore** [comment]
Okay, I was able to reproduce the problem temporarily by bringing up a fresh EKS cluster on 1.13, installing Cilium and a simple rest service, then triggering an upgrade to 1.14. Once the upgrade was complete, I deleted all of the pods for the rest service, which then got stuck with a "ContainerCreating" status for ~20 minutes. After ~20 minutes, one of the cilium pods restarted on its own, then all of the pods were able to successfully start.

Running kubectl describe pod on all of the stuck pods showed an event like:

Failed create pod sandbox: rpc error: code = Unknown desc = failed to set up sandbox container "f004ce8e1dc4536c62879409288304601d961a66286d56ef0de0ae2e9950107a" network for pod "rest-service-5f6b45b94b-cfx86": NetworkPlugin cni failed to set up pod "rest-service-5f6b45b94b-cfx86_default" network: unable to create endpoint: Put http:///var/run/cilium/cilium.sock/v1/endpoint/cilium-local:0: context deadline exceeded

I enabled debug logs while this was happening and didn't see anymore `subsys=crd-allocator` logs, or anything from the k8s client library around watches. I did run the cilium bugtool on the cilium pods while the pods were stuck - let me know if I can look for anything specific in that output.

**Comment by tom-hadlaw-hs** [comment]
Update: I've spent a bit of time digging into this issue (at least the one that we can reproduce using the previously mentioned steps). My approach was basically to add a bunch of log statements to the 1.6 branch of Cilium, build/deploy it onto a test cluster we had running and then forcing an upgrade.

Basically it seems like API calls to the Kubernetes API are hanging indefinitely. The problem arises from a call to `PUT /v1/endpoint/...`. The handler for this API call always hangs at the first call to the Kubernetes api server (in this case it's on a get Pod call).

Initially I tried adding a Timeout of 60 seconds (< cilium-cni which uses a timeout of 90 seconds). But that didn't seem to help.

Digging down further it seems like the underlying client-go kubernetes client sends the HTTP request to the apiserver but never receives a response. The read eventually times out but it seems to just keep retrying. Occasionally I would also see "connection reset" errors. This would indicate to me that this is an issue in the underlying transport, hence probably why adding a timeout to the http client doesn't do anything. I also noticed that during an EKS upgrade the Kubernetes service endpoints change indicating that they're likely changing out load-balancers behind the scenes. 

Finally in our test case I began to notice that the issue would always resolve after around 15 minutes. Eventually I found this thread: https://github.com/kubernetes/client-go/issues/374, in which people are saying that stale network connections can cause issues if there are network changes while using the Golang net transport implementations (someone mentioned that the implementation of Amazons NLBs seems to cause these issues, in particular).

To test this I changed `tcp_retries2` from 15 to 5 on all my test nodes (this corresponds roughly to how long to try exponential backoff retransmit on open connections). After rerunning the test with `tcp_retries2=5` the issue began to resolve itself after 5 minutes. 

So it seems like this issue is being called by broken tcp connections being used until the retransmit backoff runs out and the kernel closes the connection.

**However**, the one thing I'm not sure of if is whether this issue is the same issue as originally described in this thread. The big differences between our test case and what we saw on our actual cluster is:

- Issues on those clusters took much longer than 15 minutes to resolve themselves, one of our clusters remained broken for a couple days, as I remember.

- Although the root error message of PUT requests to the new endpoints endpoint timing out is the same, we observed different error messages in the logs such as `StorageError: invalid object, Code: 4` which comes from validation issues in kubeapiserver. These could be a red herring, however this was the only time we saw those errors.

Due to the circumstances of how this issue happened it seems possible that they could at least be related.

**Comment by tom-hadlaw-hs** [comment]
I should also add, our test clusters where basically vanilla eks clusters (as provisioned through Amazons Cloudformation Stack), the only workloads they had running was a deployment of nginx. I'm thinking maybe it's possible that this issue manifests itself different given larger/more complex clusters.

**Comment by tom-hadlaw-hs** [comment]
Update: Spent a bit more time trying to reproduce the `StorageError: invalid object, Code: 4` error. While reproducing the networking issue after cluster upgrade, I continually deleted all cilium identity resources. My thinking was that if this issue is related to stale cache data that maybe deleting resources while cilium was experiencing networking issues could cause the kube-apiserver validation issues we saw. 

After trying it a few times I started seeing the `StorageError: invalid object, Code: 4` error in Ciliums logs. However, as before, after 15 minutes the connections would timeout and most issues would resolve themselves. Interestingly after the issues resolved, I would still see some pods taking ~60 seconds to start due to API timeouts in Cilium (however the call would eventually succeed and the pod would be scheduled).

**Comment by tgraf** [comment]
We'll need a custom dialer similar to how kubelet has done this:
https://github.com/kubernetes/kubernetes/pull/78016/commits/1d07d303059ef3b2378cec1548f2611ec7022c5d

**Comment by tom-hadlaw-hs** [comment]
@tgraf If possible, we wouldn't mind working on this one.

**Comment by tom-hadlaw-hs** [comment]
I notice that Cilium uses a forked version of client-go, however that version does have `util/connrotation` implemented.

**Comment by tgraf** [comment]
@tom-hadlaw-hs Sure, consult with @aanm.

**Comment by aanm** [comment]
> I notice that Cilium uses a forked version of client-go, however that version does have `util/connrotation` implemented.

@tom-hadlaw-hs that forked version is the same as the upstream one with 2 more patches that are yet to be upstream. Use this branch https://github.com/cilium/client-go/tree/kubernetes-1.17.1-with-hot-fix

**Comment by tom-hadlaw-hs** [comment]
@aanm @tgraf Thanks, I've had a chance to take a look at how Kubelet does it. I'm thinking of just following what they do of using the custom dialer and then heartbeating apiserver from a goroutine.

**Comment by huntermassey** [comment]
Chiming in to say that we just recently saw the same thing this week with `Unable update CRD identity information with a reference for this node` as described in issue. 

Confirmed with AWS via a support ticket that a cluster update happened behind the scenes.

**Comment by diversario** [comment]
Saw this with 1.6.4 on a self-hosted Kops cluster after masters were rolled – some cilium pods went into this state and never recovered, causing "random" connectivity issues.

**Comment by joestringer** [comment]
@diversario the fix at this point has been merged into the master development branch for v1.8, but we'd like to be cautious on backporting to v1.7 or v1.6 branches due to the invasiveness of the change. If you're willing to help us to test this out then please reach out, we can co-ordinate on Slack.
